docs/source/notes/modules.rst

```diff
@@ -0,0 +1,433 @@
+  simple construction of complex, multi-layer neural networks.
+* **Tightly integrated with PyTorch's**
+  `autograd <https://pytorch.org/tutorials/beginner/blitz/autograd_tutorial.html>`_
+  **system.** Modules make it simple to specify learnable parameters for PyTorch's optimizers to update.
+* **Easy to work with and transform.** Modules are straightforward to save and restore, transfer between
+  CPU / GPU / TPU devices, prune, quantize, and more.
+
+This note describes modules, and is intended for all PyTorch users. Since modules are so fundamental to PyTorch,
+many topics in this note are elaborated on in other notes or tutorials, and links to many of those documents
+are provided here as well.
+
+.. contents:: :local:
+
+A Simple Custom Module
+----------------------
+
+To get started, let's look at a simpler, custom version of PyTorch's :class:`~torch.nn.Linear` module.
+This module applies an affine transformation to its input.
+
+.. code-block:: python
+
+   import torch
+   from torch import nn
+
+   class MyLinear(nn.Module):
+     def __init__(self, in_features, out_features):
+       super().__init__()
+       self.weight = nn.Parameter(torch.randn(in_features, out_features))
+       self.bias = nn.Parameter(torch.randn(out_features))
+
+     def forward(self, input):
+       return (input @ self.weight) + self.bias
+
+This simple module has the following fundamental characteristics of modules:
+
+* **It inherits from the base Module class.**
+  All modules should subclass :class:`~torch.nn.Module` for composability with other modules.
+* **It defines some "state" that is used in computation.**
+  Here, the state consists of randomly-initialized ``weight`` and ``bias`` tensors that define the affine
+  transformation. Because each of these is defined as a :class:`~torch.nn.parameter.Parameter`, the module
+  will automatically track them and return them from calls to :func:`~torch.nn.Module.parameters`. Parameters can be
+  considered the "learnable" aspects of the module's computation (more on this later). Note that modules
+  are not *required* to have state, and can also be stateless.
+* **It defines a forward() function that performs the computation.** For this affine transformation module, the input
+  is matrix-multiplied with the ``weight`` parameter (using the ``@`` short-hand notation) and added to the ``bias``
+  parameter to produce the output. More generally, the ``forward()`` implementation for a module can perform arbitrary
+  computation involving any number of inputs and outputs.
+
+This simple module demonstrates how modules package state and computation. Instances of this module can be
+constructed and called:
+
+.. code-block:: python
+
+   m = MyLinear(4, 3)
+   sample_input = torch.randn(4)
+   m(sample_input)
+   : tensor([-0.3037, -1.0413, -4.2057], grad_fn=<AddBackward0>)
+
+Note that the module itself is callable, and that calling it calls its ``forward()`` function.
```

Comment:
that's a lot of calls, maybe change the last one to invokes?

Reply:
Also, do you think it makes sense to explain why we have a forward function instead of just defining `__call__` or whatever.

Reply:
Good point- I like that idea and think it's important to cover this early on, since it's fundamental.

I suppose the terminology comes from the use of "backpropagation" as the standard for training neural nets. From this, we get the natural inverses of "forward" / "backward" passes. I added some text here briefly describing these and a link to the "Neural Network Training with Modules" section for more detail.